phoenixconnect/integrationpluginphoenixconnect.cpp

```diff
@@ -0,0 +1,254 @@
+#include <plugintimer.h>
+
+#include <QDebug>
+#include <QStringList>
+#include <QJsonDocument>
+#include <QNetworkInterface>
+
+
+IntegrationPluginPhoenixConnect::IntegrationPluginPhoenixConnect()
+{
+
+}
+
+void IntegrationPluginPhoenixConnect::discoverThings(ThingDiscoveryInfo *info)
+{
+    if (!hardwareManager()->networkDeviceDiscovery()->available()) {
+        qCWarning(dcPhoenixContact()) << "Failed to discover network devices. The network device discovery is not available.";
+        info->finish(Thing::ThingErrorHardwareNotAvailable, QT_TR_NOOP("The network cannot be searched."));
+        return;
+    }
+
+    qCDebug(dcPhoenixContact()) << "Starting network discovery...";
+    NetworkDeviceDiscoveryReply *discoveryReply = hardwareManager()->networkDeviceDiscovery()->discover();
+    connect(discoveryReply, &NetworkDeviceDiscoveryReply::finished, info, [=](){
+        ThingDescriptors descriptors;
+        qCDebug(dcPhoenixContact()) << "Discovery finished. Found" << discoveryReply->networkDeviceInfos().count() << "devices";
+        foreach (const NetworkDeviceInfo &networkDeviceInfo, discoveryReply->networkDeviceInfos()) {
+            qCDebug(dcPhoenixContact()) << networkDeviceInfo;
+
+            if (networkDeviceInfo.macAddressManufacturer() != "wallbe GmbH" && !networkDeviceInfo.macAddressManufacturer().contains("Phoenix", Qt::CaseSensitivity::CaseInsensitive)) {
+                continue;
+            }
+
+            ThingClass thingClass = supportedThings().findById(info->thingClassId());
+            ParamTypeId macAddressParamType = thingClass.paramTypes().findByName("mac").id();
+
+            ThingDescriptor descriptor(info->thingClassId(), thingClass.displayName(), networkDeviceInfo.address().toString());
+            descriptor.setParams({Param(macAddressParamType, networkDeviceInfo.macAddress())});
+
+            // Check if we already have set up this device
+            Thing *existingThing = myThings().findByParams(descriptor.params());
+            if (existingThing) {
+                qCDebug(dcPhoenixContact()) << "Found already existing" << thingClass.name() << "wallbox:" << existingThing->name() << networkDeviceInfo;
+                descriptor.setThingId(existingThing->id());
+            } else {
+                qCDebug(dcPhoenixContact()) << "Found new" << thingClass.name() << "wallbox";
+            }
+
+            info->addThingDescriptor(descriptor);
+        }
+        info->finish(Thing::ThingErrorNoError);
+    });
+}
+
+
+void IntegrationPluginPhoenixConnect::setupThing(ThingSetupInfo *info)
+{
+    Thing *thing = info->thing();
+
+    if (m_connections.contains(thing)) {
+        qCDebug(dcPhoenixContact()) << "Reconfiguring existing thing" << thing->name();
+        m_connections.take(thing)->deleteLater();
+    } else {
+        qCDebug(dcPhoenixContact()) << "Setting up a new device:" << thing->params();
+    }
+
+
+    QString mac = thing->paramValue("mac").toString();
```

Comment:
Since you allow also a setup method `user` you should probably check if the given mac address is valid `MacAddress::isValid()`